### ISSUE TITLE
Bounding Boxes displacement error

### ISSUE THREAD

**Issue by pecia6** [aditya-vora/FCHD-Fully-Convolutional-Head-Detector#15]
I have run "head_detection_demo.py" and I haven't received good results.
The bounding boxes seem correct but there is a kind of displacement respect to ground truth.

![prova2](https://user-images.githubusercontent.com/36638318/48846008-47cc7f00-ed9e-11e8-842c-68b0fc977e33.png)

**Comment by SmartGx** [comment]
I got the same problem as the above picture, and got some errors about `urllib3`.
Can you tell me if you solve this problem now?

**Comment by AngelaDevHao** [comment]
Have you already fixed this problem? It seems i got the same one, and could you tell me please?

**Comment by dagongji10** [comment]
> I got the same problem as the above picture, and got some errors about `urllib3`.
> Can you tell me if you solve this problem now?

I have got the same issue with you sadly.
Do you solve the problem now? I really don't understand why 'urllib3' is also throw error.

**Comment by okideal** [comment]
change line 57 to
utils.draw_bounding_box_on_image_array(img_raw, ymin/scale, xmin/scale, ymax/scale, xmax/scale)

**Comment by dagongji10** [comment]
@SmartGx @pecia6 I know why this happen now.There are 3 places need to change in head_detection_demo.py.
(1) IM_RESIZE=True
(2)place the code ‘_, H, W = img.shape’ after 'img = img.transpose((2,0,1))', because later we will use H to  calculate scale, but before transpose the img is (h, w, c)
(3)change the code 'f = f.resize((640,480), Image.ANTIALIAS)' to 'f = f.resize((800,600), Image.ANTIALIAS)'. this is because the author draw box on img_raw which is (640, 480), while he get the box with img which is (800, 600)

**Comment by dagongji10** [comment]
> change line 57 to
> utils.draw_bounding_box_on_image_array(img_raw, ymin/scale, xmin/scale, ymax/scale, xmax/scale)

also need to place ‘_, H, W = img.shape’  after  ‘img = img.transpose((2,0,1))’

**Comment by okideal** [comment]
This work is great but the code is absolutely disaster

**Comment by aditya-vora** [comment]
Hi Guys, 
I am sorry you are facing these issue of displacement error. However, I didn't face any such issue when I tested the model on this dataset. However, I need to debug the code once. Give me some time I will rectify the error and get back soon.

**Comment by dagongji10** [comment]
> Hi Guys,
> I am sorry you are facing these issue of displacement error. However, I didn't face any such issue when I tested the model on this dataset. However, I need to debug the code once. Give me some time I will rectify the error and get back soon.

hei, brother, you really did a great work. I can run the test-demo now and get a nice result. Thanks for your work very much.

**Comment by aditya-vora** [comment]
@dagongji10 Glad that you liked the work. May I know what changes you made in order to make it work?

**Comment by dagongji10** [comment]
> @dagongji10 Glad that you liked the work. May I know what changes you made in order to make it work?

I think maybe the code ‘_, H, W = img.shape’  should after  ‘img = img.transpose((2,0,1))’， so we can get the right H and scale.
The box predict is very accurate，but when draw it on img_raw , maybe we can change like that:
'ymin, xmin, ymax, xmax = pred_bboxes_[i,:]/scale'
then the box can align to img_raw.

I don't konw whether am i right, but it really works for me.

**Comment by mrlonely001** [comment]
> > change line 57 to
> > utils.draw_bounding_box_on_image_array(img_raw, ymin/scale, xmin/scale, ymax/scale, xmax/scale)
> 
> also need to place ‘_, H, W = img.shape’ after ‘img = img.transpose((2,0,1))’

thanks bro, nice job, and @aditya-vora nice project ~

**Comment by pecia6** [comment]
Hi Guys,
I confirm that your fix solves the displacement problem, but it seems to make the performance worse.

For example, the image I posted previously becomes this:
 
![prova3](https://user-images.githubusercontent.com/36638318/50760313-af179300-1267-11e9-8518-865c3a7678ae.png)

There are four false negative more and a false positive more.

**Comment by x2ss** [comment]
Hi guys,

I got no head detections, just like [Issue#3](https://github.com/aditya-vora/FCHD-Fully-Convolutional-Head-Detector/issues/3)(closed by author).

Also, after change line 57 to utils.draw_bounding_box_on_image_array(img_raw, ymin/scale, xmin/scale, ymax/scale, xmax/scale), I got blank picture just like this:

![figure_1](https://user-images.githubusercontent.com/37960611/51926306-f559af80-242b-11e9-9c0e-e8eb9d61ded7.png)


As you have run successfully, would you please tell me:

1. Which version of cuda and pytorch did you use?

2. What changes needed to be added to the config.py?

3. Why did I get blank picture? my code after change is 
      for i in range(pred_bboxes_.shape[0]):
        ymin, xmin, ymax, xmax = pred_bboxes_[i,:]
        utils.draw_bounding_box_on_image_array(img_raw,ymin, xmin, ymax, xmax)
    plt.axis('off')
    utils.draw_bounding_box_on_image_array(img_raw, ymin / scale, xmin / scale, ymax / scale, xmax / scale)
    if SAVE_FLAG == 1:
        plt.savefig(os.path.join(opt.test_output_path, file_id+'.png'), bbox_inches='tight', pad_inches=0)
    else:
        plt.show()    

Thank you very much. Your kind replay would be highly appreciated.

**Comment by sagrawal1993** [comment]
@pecia6 I am facing same problem with the accuracy. Have you found any solution?

**Comment by sunjunlishi** [comment]
i am facing same problem,Rect not ok

**Comment by sunjunlishi** [comment]
![out](https://user-images.githubusercontent.com/11845664/58160211-9b1d7c80-7cb0-11e9-82c3-fc89051dad06.png)
![test](https://user-images.githubusercontent.com/11845664/58160213-9ce74000-7cb0-11e9-8f99-d36877d45811.jpg)

**Comment by sahilrider** [comment]
> ![out](https://user-images.githubusercontent.com/11845664/58160211-9b1d7c80-7cb0-11e9-82c3-fc89051dad06.png)
> ![test](https://user-images.githubusercontent.com/11845664/58160213-9ce74000-7cb0-11e9-8f99-d36877d45811.jpg)

I tested this code on your image and I am also getting the same results. Definitely not OK!
Did you find any other better methods for head detector?

**Comment by FarisHijazi** [comment]
Could somebody please share the weights of the pretrained model? it seems to have been taken down
You can email me at [theefaris@gmail.com](mailto:theefaris@gmail.com) if you have the weights, thank you!
